tests/DependencyInjection/Compiler/RegisterMenusPassTest.php

```diff
@@ -21,93 +22,93 @@ class RegisterMenusPassTest extends TestCase
+        $this->containerBuilder->getDefinition('id')
+            ->setTags(['knp_menu.menu_builder' => [['alias' => 'foo']]]);
 
         $this->expectException(\InvalidArgumentException::class);
         $this->expectExceptionMessage('The method is not defined in the "knp_menu.menu_builder" tag for the service "id"');
-        $this->pass->process($this->containerBuilder->reveal());
+
+        $this->pass->process($this->containerBuilder);
     }
 
     public function testFailsWhenMenuAliasIsMissing(): void
     {
-        $this->containerBuilder->findTaggedServiceIds('knp_menu.menu', true)->willReturn(['id' => [['alias' => '']]]);
+        $this->containerBuilder->getDefinition('menu_id')
+            ->setTags(['knp_menu.menu' => [['alias' => '']]]);
 
         $this->expectException(\InvalidArgumentException::class);
-        $this->expectExceptionMessage('The alias is not defined in the "knp_menu.menu" tag for the service "id"');
-        $this->pass->process($this->containerBuilder->reveal());
+        $this->expectExceptionMessage('The alias is not defined in the "knp_menu.menu" tag for the service "menu_id"');
+
+        $this->pass->process($this->containerBuilder);
     }
 
     public function testRegisterMenuBuilderAndMenu(): void
     {
-        $this->containerBuilder->findTaggedServiceIds('knp_menu.menu_builder', true)->willReturn([
-            'id1' => [['alias' => 'foo', 'method' => 'fooMenu'], ['alias' => 'bar', 'method' => 'bar']],
-            'id2' => [['alias' => 'foo', 'method' => 'fooBar'], ['alias' => 'baz', 'method' => 'bar']],
-        ]);
-        $this->containerBuilder->findTaggedServiceIds('knp_menu.menu', true)->willReturn(['id3' => [['alias' => 'test']]]);
-
-        $menus = [
-            'foo' => [new ServiceClosureArgument(new Reference('id2')), 'fooBar'],
-            'bar' => [new ServiceClosureArgument(new Reference('id1')), 'bar'],
-            'baz' => [new ServiceClosureArgument(new Reference('id2')), 'bar'],
-            'test' => new ServiceClosureArgument(new Reference('id3')),
-        ];
-        $this->definition->replaceArgument(0, $menus)->shouldBeCalled();
-
-        $this->pass->process($this->containerBuilder->reveal());
+        $this->pass->process($this->containerBuilder);
+
+        $menuBuilders = $this->containerBuilder->getDefinition('knp_menu.menu_provider.lazy')->getArgument(0);
+
+        $this->assertFalse($this->containerBuilder->hasDefinition('knp_menu.menu_provider.builder_service'));
+        $this->assertArrayHasKey('foo', $menuBuilders);
+        $this->assertArrayHasKey('bar', $menuBuilders);
+        $this->assertArrayHasKey('baz', $menuBuilders);
+        $this->assertIsArray($menuBuilders['bar']);
+        $this->assertIsArray($menuBuilders['foo']);
+        $this->assertSame('id', (string) $menuBuilders['foo'][0]->getValues()[0]);
+        $this->assertSame('id', (string) $menuBuilders['bar'][0]->getValues()[0]);
```

Comment:
There is no assertion that we get a proper ServiceClosureArgument here, but an assumption that `getValues` can be called (and that would also work with an IteratorArgument without making the test fail while the code would be broken).

The equivalent of the old test is `$this->assertEquals($menus, $this->containerBuilder->getDefinition('knp_menu.menu_provider.lazy')->getArgument(0));`, with the existing `$menus` variable

Reply:
Fixed by checking with the complete array of expected results.